modules/gapi/test/serialization/serialization.cpp

```diff
@@ -0,0 +1,1292 @@
+            os << (uint)mat_desc.dims.data()[j];
+        }
+        break;
+    }
+    case cv::GShape::GSCALAR:
+    {
+        //std::cout << "dumpGSerializedDatas  GSCALAR " << std::endl;
+        //GScalarDesc scalar_desc = util::get<GScalarDesc>(data.meta);
+        break;
+    }
+    case cv::GShape::GARRAY:
+    {   //std::cout << "dumpGSerializedDatas  GARRAY " << std::endl;
+        //GArrayDesc array_desc = util::get<GArrayDesc>(data.meta);
+        break;
+    }
+    case cv::GShape::GOPAQUE:
+    {
+        //std::cout << "dumpGSerializedDatas  GOPAQUE " << std::endl;
+        //GOpaqueDesc opaque_desc = util::get<GOpaqueDesc>(data.meta);
+        break;
+    }
+    default:
+        //std::cout << "dumpGSerializedDatas  unsupported" << std::endl;
+        break;
+    }
+
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, Op op)
+{
+    //Kernel k
+    os << op.k;
+
+    //std::vector<int>   kind;
+    os << op.kind;
+
+    //std::vector<int>   opaque_kind;
+    os << op.opaque_kind;
+
+    //std::vector<RcDesc> outs;
+    os << op.outs;
+
+    //std::vector<RcDesc> ins;
+    os << op.ins;
+
+    //opaque args
+    //std::vector<int> opaque_ints;
+    os << op.opaque_ints;
+
+    //std::vector<double> opaque_doubles;
+    os << op.opaque_doubles;
+
+    //std::vector<cv::Size> opaque_cvsizes;
+    os << op.opaque_cvsizes;
+
+    //std::vector<bool> opaque_bools;
+    os << op.opaque_bools;
+
+    //std::vector<cv::Scalar> opaque_cvscalars;
+    os << op.opaque_cvscalars;
+
+    //std::vector<cv::Point> opaque_cvpoints;
+    os << op.opaque_cvpoints;
+
+    //std::vector<cv::Mat> opaque_cvmats;
+    os << op.opaque_cvmats;
+
+    //std::vector<cv::Rect> opaque_cvrects;
+    os << op.opaque_cvrects;
+
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, std::vector<Op> ops)
+{
+    //dumpAtomS(os, (uint)ops.size());
+    os << (uint)ops.size();
+    for (const auto& op : ops)
+    {
+        //Op
+        os << op;
+    }
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, int atom)
+{
+    dumpAtom(os, atom);
+    return os;
+}
+
+SerializationStream& operator << (SerializationStream& os, uint atom)
+{
+    dumpAtom(os, atom);
+    return os;
+}
+
+void dumpAtom(SerializationStream &os, uint atom)
+{
+    uint atom_htonl = htonl(atom);
+    os.putAtom(atom_htonl);
+}
+
+void dumpGSerialized(const GSerialized s, SerializationStream &ofs_serialized)
+{
+    ofs_serialized << s.m_ops;
+    ofs_serialized << s.m_datas;
+}
+
+//Graph restore operators
+template<typename _Tp> static inline
+DeSerializationStream& operator >> (DeSerializationStream& is, /*const*/ _Tp& value)
+{
+    value >> is;
+    return is;
+}
+
+DeSerializationStream& operator >> (DeSerializationStream& is, Kernel& k)
+{
+    is >> k.name;
+    is >> k.tag;
+    CV_LOG_INFO(NULL, "k.name  " << k.name.c_str());
+    CV_LOG_INFO(NULL, "k.tag  " << k.tag.c_str());
+    return is;
+}
```

Comment:
The right approach in writing serialization is putting `<<` and `>>` for every type nearby. That is, every your `<<` needs to have `>>` right below. Otherwise it is easy to make a mistake and hard to maintain the serialization integrity.